### PR TITLE
[SPARK-26352][SQL] join reorder should not change the order of output attributes

### PR DESCRIPTION
**Pull request by rednaxelafx**
## What changes were proposed in this pull request?

The optimizer rule `org.apache.spark.sql.catalyst.optimizer.ReorderJoin` performs join reordering on inner joins. This was introduced from SPARK-12032 (https://github.com/apache/spark/pull/10073) in 2015-12.

After it had reordered the joins, though, it didn't check whether or not the output attribute order is still the same as before. Thus, it's possible to have a mismatch between the reordered output attributes order vs the schema that a DataFrame thinks it has.
The same problem exists in the CBO version of join reordering (`CostBasedJoinReorder`) too.

This can be demonstrated with the example:
```scala
spark.sql("create table table_a (x int, y int) using parquet")
spark.sql("create table table_b (i int, j int) using parquet")
spark.sql("create table table_c (a int, b int) using parquet")
val df = spark.sql("""
  with df1 as (select * from table_a cross join table_b)
  select * from df1 join table_c on a = x and b = i
""")
```
here's what the DataFrame thinks:
```
scala> df.printSchema
root
 |-- x: integer (nullable = true)
 |-- y: integer (nullable = true)
 |-- i: integer (nullable = true)
 |-- j: integer (nullable = true)
 |-- a: integer (nullable = true)
 |-- b: integer (nullable = true)
```
here's what the optimized plan thinks, after join reordering:
```
scala> df.queryExecution.optimizedPlan.output.foreach(a => println(s"|-- ${a.name}: ${a.dataType.typeName}"))
|-- x: integer
|-- y: integer
|-- a: integer
|-- b: integer
|-- i: integer
|-- j: integer
```

If we exclude the `ReorderJoin` rule (using Spark 2.4's optimizer rule exclusion feature), it's back to normal:
```
scala> spark.conf.set("spark.sql.optimizer.excludedRules", "org.apache.spark.sql.catalyst.optimizer.ReorderJoin")

scala> val df = spark.sql("with df1 as (select * from table_a cross join table_b) select * from df1 join table_c on a = x and b = i")
df: org.apache.spark.sql.DataFrame = [x: int, y: int ... 4 more fields]

scala> df.queryExecution.optimizedPlan.output.foreach(a => println(s"|-- ${a.name}: ${a.dataType.typeName}"))
|-- x: integer
|-- y: integer
|-- i: integer
|-- j: integer
|-- a: integer
|-- b: integer
```

Note that this output attribute ordering problem leads to data corruption, and can manifest itself in various symptoms:
* Silently corrupting data, if the reordered columns happen to either have matching types or have sufficiently-compatible types (e.g. all fixed length primitive types are considered as "sufficiently compatible" in an `UnsafeRow`), then only the resulting data is going to be wrong but it might not trigger any alarms immediately. Or
* Weird Java-level exceptions like `java.lang.NegativeArraySizeException`, or even SIGSEGVs.

## How was this patch tested?

Added new unit test in `JoinReorderSuite` and new end-to-end test in `JoinSuite`.
Also made `JoinReorderSuite` and `StarJoinReorderSuite` assert more strongly on maintaining output attribute order.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 1]
```diff
@@ -895,4 +895,18 @@ class JoinSuite extends QueryTest with SharedSQLContext {
...
+    withTable("tab1", "tab2", "tab3") {
+      spark.sql("select 1 as x, 100 as y").write.saveAsTable("tab1")
+      spark.sql("select 42 as i, 200 as j").write.saveAsTable("tab2")
+      spark.sql("select 1 as a, 42 as b").write.saveAsTable("tab3")
+
+      val df = spark.sql("""
+        with tmp as (select * from tab1 cross join tab2)
+        select * from tmp join tab3 on a = x and b = i
+      """)
+      checkAnswer(df, Row(1, 100, 42, 200, 1, 42))
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]:
```Scala
scala> spark.sql("with tmp as (select * from tab1 cross join tab2) select * from tmp join tab3 on a = x and b = i").show()
+---+---+---+---+---+---+
|  x|  y|  i|  j|  a|  b|
+---+---+---+---+---+---+
|  1|100| 42|200|  1| 42|
+---+---+---+---+---+---+
```

This still passes without code changes, right?

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: Maybe the command lines work but the test case does not work since we turn off `ConvertToLocalRelation` in the test environment?

Comment (rednaxelafx) [reply sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: Yeah, before sending this PR I've actually explicitly run both test cases that I added with and without the fix, and verified that without the fix it'd fail the test.

The reason why you "think" it works in your environment is because you used `show`, which after @maropu 's earlier PR will introduce a `Project` at the top, which happens to cure the issue.
If you just do a `collect` you'll see the problem still exists:
```
scala> df.show
+---+---+---+---+---+---+
|  x|  y|  i|  j|  a|  b|
+---+---+---+---+---+---+
|  1|100| 42|200|  1| 42|
+---+---+---+---+---+---+

scala> df.collect
res5: Array[org.apache.spark.sql.Row] = Array([1,100,1,42,42,200])
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala [hunk 2]
```diff
@@ -99,6 +99,14 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
       } else {
         createOrderedJoin(input, conditions)
       }
+
+      if (p.sameOutput(reordered)) {
+        reordered
+      } else {
+        // Reordering the joins have changed the order of the columns.
+        // Inject a projection to make sure we restore to the expected ordering.
+        Project(p.output, reordered)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala]: just for curiosity, we only need this for top-level join? I feel it's ok to change the columnar order for intermedia joins.

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 3]
```diff
@@ -895,4 +895,18 @@ class JoinSuite extends QueryTest with SharedSQLContext {
       checkAnswer(res, Row(0, 0, 0))
     }
   }
+
+  test("SPARK-26352: join reordering should not change the order of columns") {
+    withTable("tab1", "tab2", "tab3") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: can we reproduce the bug with temp view?

**Comment by cloud-fan** [comment]
good catch! LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala (hunk 3 above)
Comment (rednaxelafx) [reply sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: That's more prone to be subject to `ConvertToLocalRelation`, i.e. you can't copy-and-paste the test case into a Spark shell and expect the same behavior. Would you still prefer to use temp view instead?

**Comment by rednaxelafx** [comment]
@cloud-fan @gatorsmile @hvanhovell I've updated the PR with a fix to the same problem in CBO join reorder, and also fixed some test cases that have been wrong for a long time (the `expected` plans were in fact wrong).

PTAL, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 4]
```diff
@@ -48,8 +48,18 @@ object CostBasedJoinReorder extends Rule[LogicalPlan] with PredicateHelper {
           if projectList.forall(_.isInstanceOf[Attribute]) =>
           reorder(p, p.output)
       }
-      // After reordering is finished, convert OrderedJoin back to Join
+
+      // Cleanups
       result transformDown {
+        // if a Project was created to keep output attribute order after join reordering, but
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: is this an improvement? Maybe we should do it in another PR, so that this PR can be backported to old branches.

Comment (rednaxelafx) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: This isn't really an improvement. It's here mainly to help pass existing tests -- I'm adding new projections to fix the output attribute order problem, but these extra projections in the middle would make (`expected`) query plans look pretty ugly.

So consider this is two things that can cancel each other out:
1. Add projections to fix output attribute order;
2. If one of these extra projections is in the middle, get rid of it.

The cleanup (2) is only meant to clean extra projections created because of (1). Both (1) and (2) are in this PR so I don't consider this as a performance improvement over existing stuff.

**Comment by gatorsmile** [comment]
@rednaxelafx Can we separate this to two PRs? One is for the non-CBO reordering. Another is for the CBO reordering?

**Comment by rednaxelafx** [comment]
@gatorsmile that was my original intent. I only wanted to fix the non-CBO one in this PR originally, but after reading through the join reorder related test suite I found that most of them have both the non-CBO and the CBO one in the same `Optimize`. The test cases are going to fail one way or the other if we only fix one of them. Sadly we'll have to fix both in the same PR.

The feature fix itself is really small, but granted it's in a very dense piece of code. The majority of this PR is actually about fixing the test cases....a lot of them were wrong to begin with (had incorrect output attribute order in the `expected` plans)

**Comment by rednaxelafx** [comment]
jenkins retest this please

**Comment by rednaxelafx** [comment]
jenkins retest this please

**Comment by rednaxelafx** [comment]
As per offline discussion with @cloud-fan and @hvanhovell , I've updated the PR and moved the logic that eliminates redundant projections from the CBO-based join reorder to `RemoveRedundantProject`.

Could you please take another look? Thanks!
cc @gatorsmile

**Comment by rednaxelafx** [comment]
Oops, the relaxed transformation in `RemoveRedundantProject` is too wide and it's wrong now. Finding alternatives now.

**Comment by rednaxelafx** [comment]
jenkind retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/StarJoinReorderSuite.scala [hunk 5]
```diff
@@ -395,6 +398,7 @@ class StarJoinReorderSuite extends PlanTest with StatsEstimationTestBase {
         .join(d2.where(nameToAttr("d2_c2") === 2), Inner,
           Some(nameToAttr("f1_fk2") === nameToAttr("d2_pk1")))
         .join(s3, Inner, Some(nameToAttr("f11_fk1") === nameToAttr("s3_pk1")))
+        .select(outputsOf(d1, f11, f1, d2, s3): _*)
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/StarJoinReorderSuite.scala]: Could you help update the comment in line 384?

**Comment by cloud-fan** [comment]
thanks, merging to master/2.4/2.3!
